grails-app/controllers/org/transmartproject/rest/HighDimController.groovy

```diff
@@ -48,39 +54,82 @@ class HighDimController {
     @Resource
     StudyLoadingService studyLoadingServiceProxy
 
-    def show(String dataType, String projection) {
-
-        if (dataType == null) {
-            index()
-        } else {
-            exportData(dataType, projection)
+    def index() {
+        if (params.dataType) {
+            // backwards compatibility
+            // preferred to use /highdim/<data type> for download
+            download params.dataType
+            return
         }
+
+        String conceptKey = getConceptKey(params.conceptId)
+        OntologyTerm concept = conceptsResourceService.getByKey(conceptKey)
+        String conceptLink = studyLoadingServiceProxy.getOntologyTermUrl(concept)
+        String selfLink = HighDimSummarySerializationHelper.getHighDimIndexUrl(conceptLink)
+
+        respond wrapList(getHighDimSummaries(concept), selfLink)
     }
 
-    private void exportData(String dataType, String projection) {
+    def download(String dataType) {
+        assert dataType != null // ensured by mapping
+        def assayConstraintsSpec = processConstraintsJson params.assayConstraints
+        def dataConstraintsSpec = processConstraintsJson params.dataConstraints
+
         String conceptKey = getConceptKey(params.conceptId)
-        OutputStream out = response.outputStream
-        response.contentType =  'application/octet-stream'
+        OutputStream out = new LazyOutputStreamDecorator(
+                outputStreamProducer: { ->
+                    response.contentType = 'application/octet-stream'
+                    response.outputStream
+                })
 
         try {
-            highDimDataService.write(conceptKey, dataType, projection, out)
-            out.flush()
+            highDimDataService.write(conceptKey, dataType, params.projection,
+                    assayConstraintsSpec, dataConstraintsSpec, out)
         } finally {
             out.close()
         }
     }
 
-    private String getConceptKey(String concept) {
-        OntologyTermCategory.keyFromURLPart(concept, studyLoadingServiceProxy.study)
-    }
+    private Map<String, List> processConstraintsJson(String paramValue) {
```

Comment:
Why wouldn't it belong here? All this is doing is parsing is parsing the client input and deserializing it into a form to pass to the service. This seems to be a function appropriate for the controller, at least absent the use of the data binding functionality of Grails. But certainly a "utility class" is not the solution, utility classes are provide no useful cohesion.